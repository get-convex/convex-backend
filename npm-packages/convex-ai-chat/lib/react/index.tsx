import {
  ConvexProvider,
  ConvexReactClient,
  useMutation,
  useQuery,
} from "convex/react";
import {
  FormEvent,
  ReactNode,
  useCallback,
  useEffect,
  useMemo,
  useRef,
  useState,
} from "react";
import { createPortal } from "react-dom";
import { api } from "../../convex/_generated/api.js";
import { CloseIcon } from "./CloseIcon.js";
import { InfoCircled } from "./InfoCircled.js";
import { SendIcon } from "./SendIcon.js";
import { SizeIcon } from "./SizeIcon.js";
import { TrashIcon } from "./TrashIcon.js";

export function ConvexAiChat({
  convexUrl,
  infoMessage,
  welcomeMessage,
  renderTrigger,
}: {
  convexUrl: string;
  infoMessage: ReactNode;
  welcomeMessage: string;
  renderTrigger: (onClick: () => void) => ReactNode;
}) {
  const [hasOpened, setHasOpened] = useState(false);
  const [dialogOpen, setDialogOpen] = useState(false);

  const handleCloseDialog = useCallback(() => {
    setDialogOpen(false);
  }, []);

  return (
    <>
      {renderTrigger(() => {
        setHasOpened(true);
        setDialogOpen(!dialogOpen);
      })}
      {hasOpened
        ? createPortal(
            <ConvexAiChatDialog
              convexUrl={convexUrl}
              infoMessage={infoMessage}
              isOpen={dialogOpen}
              welcomeMessage={welcomeMessage}
              onClose={handleCloseDialog}
            />,
            document.body,
          )
        : null}
    </>
  );
}

export function ConvexAiChatDialog({
  convexUrl,
  infoMessage,
  isOpen,
  welcomeMessage,
  onClose,
}: {
  convexUrl: string;
  infoMessage: ReactNode;
  isOpen: boolean;
  welcomeMessage: string;
  onClose: () => void;
}) {
  const client = useMemo(() => new ConvexReactClient(convexUrl), [convexUrl]);

  return (
    <ConvexProvider client={client}>
      <Dialog
        infoMessage={infoMessage}
        isOpen={isOpen}
        welcomeMessage={welcomeMessage}
        onClose={onClose}
      />
    </ConvexProvider>
  );
}

export function Dialog({
  infoMessage,
  isOpen,
  welcomeMessage,
  onClose,
}: {
  infoMessage: ReactNode;
  isOpen: boolean;
  welcomeMessage: string;
  onClose: () => void;
}) {
  const [sessionId, resetSessionId] = useSessionId();
  const remoteMessages = useQuery(api.messages.list, { sessionId });
  const messages = useMemo(
    () =>
      [{ isViewer: false, text: welcomeMessage, _id: "0" }].concat(
        (remoteMessages ?? []) as {
          isViewer: boolean;
          text: string;
          _id: string;
        }[],
      ),
    [remoteMessages, welcomeMessage],
  );
  const sendMessage = useMutation(api.messages.send);

  const [expanded, setExpanded] = useState(false);
  const [isScrolled, setScrolled] = useState(false);

  const [input, setInput] = useState("");

  const handleExpand = () => {
    setExpanded(!expanded);
    setScrolled(false);
  };

  const handleSend = async (event: FormEvent) => {
    event.preventDefault();
    await sendMessage({ message: input, sessionId });
    setInput("");
    setScrolled(false);
  };

  const handleClearMessages = async () => {
    resetSessionId();
    setScrolled(false);
  };

  const listRef = useRef<HTMLDivElement>(null);

  useEffect(() => {
    if (isScrolled) {
      return;
    }
    // Using `setTimeout` to make sure scrollTo works on button click in Chrome
    setTimeout(() => {
      listRef.current?.scrollTo({
        top: listRef.current.scrollHeight,
        behavior: "smooth",
      });
    }, 0);
  }, [messages, isScrolled]);

  return (
    <div
      id="convex-ai-chat"
      className={
        (isOpen ? "fixed" : "hidden") +
        " rounded-xl flex flex-col bg-white dark:bg-black text-black dark:text-white " +
        "m-4 right-0 bottom-0 max-w-[calc(100%-2rem)] overflow-hidden transition-all " +
        "shadow-[0px_5px_40px_rgba(0,0,0,0.16),0_20px_25px_-5px_rgb(0,0,0,0.1)] " +
        "dark:shadow-[0px_5px_40px_rgba(0,0,0,0.36),0_20px_25px_-5px_rgb(0,0,0,0.3)] " +
        (expanded
          ? "left-0 top-0 z-[1000]"
          : "w-full sm:max-w-[25rem] sm:min-w-[25rem] h-[30rem]")
      }
    >
      <div className="flex justify-end">
        <button
          className="group border-none bg-transparent p-0 pt-2 px-2 cursor-pointer hover:text-neutral-500 dark:hover:text-neutral-300"
          onClick={handleClearMessages}
        >
          <InfoCircled className="h-5 w-5" />
          <span
            className={
              "invisible absolute z-50 cursor-auto group-hover:visible text-base text-black dark:text-white " +
              "rounded-md shadow-[0px_5px_12px_rgba(0,0,0,0.32)] p-2 bg-white dark:bg-neutral-700 top-12 right-8 left-8 text-center"
            }
          >
            {infoMessage}
          </span>
        </button>
        <button
          className="border-none bg-transparent p-0 pt-2 px-2 cursor-pointer hover:text-neutral-500 dark:hover:text-neutral-300"
          onClick={handleClearMessages}
        >
          <TrashIcon className="h-5 w-5" />
        </button>
        <button
          className="border-none bg-transparent p-0 pt-2 px-2 cursor-pointer hover:text-neutral-500 dark:hover:text-neutral-300"
          onClick={handleExpand}
        >
          <SizeIcon className="h-5 w-5" />
        </button>
        <button
          className="border-none bg-transparent p-0 pt-2 px-2 cursor-pointer hover:text-neutral-500 dark:hover:text-neutral-300"
          onClick={onClose}
        >
          <CloseIcon className="h-5 w-5" />
        </button>
      </div>
      <div
        className="flex-grow overflow-scroll gap-2 flex flex-col mx-2 pb-2 rounded-lg"
        ref={listRef}
        onWheel={() => {
          setScrolled(true);
        }}
      >
        {remoteMessages === undefined ? (
          <>
            <div className="animate-pulse rounded-md bg-black/10 h-5" />
            <div className="animate-pulse rounded-md bg-black/10 h-9" />
          </>
        ) : (
          messages.map((message) => (
            <div key={message._id}>
              <div
                className={
                  "text-neutral-400 text-sm " +
                  (message.isViewer && !expanded ? "text-right" : "")
                }
              >
                {message.isViewer ? <>You</> : <>Convex AI Bot</>}
              </div>
              {message.text === "" ? (
                <div className="animate-pulse rounded-md bg-black/10 h-9" />
              ) : (
                <div
                  className={
                    "w-full rounded-xl px-3 py-2 whitespace-pre-wrap " +
                    (message.isViewer
                      ? "bg-neutral-200 dark:bg-neutral-800 "
                      : "bg-neutral-100 dark:bg-neutral-900 ") +
                    (message.isViewer && !expanded
                      ? "rounded-tr-none"
                      : "rounded-tl-none")
                  }
                >
                  {message.text}
                </div>
              )}
            </div>
          ))
        )}
      </div>
      <form
        className="border-t-neutral-200 dark:border-t-neutral-800 border-solid border-0 border-t-[1px] flex"
        onSubmit={handleSend}
      >
        <input
          className="w-full bg-white dark:bg-black border-none text-[1rem] pl-4 py-3 outline-none"
          autoFocus
          name="message"
          placeholder="Send a message"
          value={input}
          onChange={(event) => setInput(event.target.value)}
        />
        <button
          disabled={input === ""}
          className="bg-transparent border-0 px-4 py-3 enabled:cursor-pointer enabled:hover:text-sky-500"
        >
          <SendIcon className="w-5 h-5" />
        </button>
      </form>
    </div>
  );
}

const STORE = (typeof window === "undefined" ? null : window)?.sessionStorage;
const STORE_KEY = "ConvexSessionId";

function useSessionId() {
  const [sessionId, setSessionId] = useState(
    () => STORE?.getItem(STORE_KEY) ?? crypto.randomUUID(),
  );

  useEffect(() => {
    STORE?.setItem(STORE_KEY, sessionId);
  }, [sessionId]);

  const resetSessionId = useCallback(() => {
    setSessionId(crypto.randomUUID());
  }, []);

  return [sessionId, resetSessionId] as const;
}
