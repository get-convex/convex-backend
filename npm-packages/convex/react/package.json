{
  "main": "../dist/cjs/react/index.js",
  "module": "../dist/esm/react/index.js",
  "types": "../dist/cjs-types/react/index.d.ts"
}
