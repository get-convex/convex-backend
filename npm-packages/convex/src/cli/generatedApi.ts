/**
 * This file was auto-generated by openapi-typescript.
 * Do not make direct changes to the file.
 */

export interface paths {
    "/teams": {
        parameters: {
            query?: never;
            header?: never;
            path?: never;
            cookie?: never;
        };
        get: operations["get_teams_for_member"];
        put?: never;
        post?: never;
        delete?: never;
        options?: never;
        head?: never;
        patch?: never;
        trace?: never;
    };
    "/teams/{team_slug}/projects": {
        parameters: {
            query?: never;
            header?: never;
            path?: never;
            cookie?: never;
        };
        get: operations["get_cli_projects_for_team"];
        put?: never;
        post?: never;
        delete?: never;
        options?: never;
        head?: never;
        patch?: never;
        trace?: never;
    };
}
export type webhooks = Record<string, never>;
export interface components {
    schemas: {
        /** Format: int64 */
        MemberId: number;
        ProjectDetails: {
            /** Format: int64 */
            createTime: number;
            id: components["schemas"]["ProjectId"];
            isDemo: boolean;
            name: components["schemas"]["ProjectName"];
            slug: components["schemas"]["ProjectSlug"];
            teamId: components["schemas"]["TeamId"];
        };
        /** Format: int64 */
        ProjectId: number;
        ProjectName: string;
        ProjectSlug: string;
        ReferralCode: string;
        Team: {
            creator?: null | components["schemas"]["MemberId"];
            id: components["schemas"]["TeamId"];
            managedBy?: string | null;
            name: components["schemas"]["TeamName"];
            referralCode: components["schemas"]["ReferralCode"];
            referredBy?: null | components["schemas"]["TeamId"];
            slug: components["schemas"]["TeamSlug"];
            suspended: boolean;
        };
        /** Format: int64 */
        TeamId: number;
        TeamName: string;
        TeamSlug: string;
    };
    responses: never;
    parameters: never;
    requestBodies: never;
    headers: never;
    pathItems: never;
}
export type MemberId = components['schemas']['MemberId'];
export type ProjectDetails = components['schemas']['ProjectDetails'];
export type ProjectId = components['schemas']['ProjectId'];
export type ProjectName = components['schemas']['ProjectName'];
export type ProjectSlug = components['schemas']['ProjectSlug'];
export type ReferralCode = components['schemas']['ReferralCode'];
export type Team = components['schemas']['Team'];
export type TeamId = components['schemas']['TeamId'];
export type TeamName = components['schemas']['TeamName'];
export type TeamSlug = components['schemas']['TeamSlug'];
export type $defs = Record<string, never>;
export interface operations {
    get_teams_for_member: {
        parameters: {
            query?: never;
            header?: never;
            path?: never;
            cookie?: never;
        };
        requestBody?: never;
        responses: {
            200: {
                headers: {
                    [name: string]: unknown;
                };
                content: {
                    "application/json": components["schemas"]["Team"][];
                };
            };
        };
    };
    get_cli_projects_for_team: {
        parameters: {
            query?: never;
            header?: never;
            path: {
                /** @description Team slug identifier */
                team_slug: string;
            };
            cookie?: never;
        };
        requestBody?: never;
        responses: {
            200: {
                headers: {
                    [name: string]: unknown;
                };
                content: {
                    "application/json": components["schemas"]["ProjectDetails"][];
                };
            };
        };
    };
}
