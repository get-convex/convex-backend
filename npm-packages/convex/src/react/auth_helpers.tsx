import React, { ReactNode } from "react";
import { useConvexAuth } from "./ConvexAuthState.js";

/**
 * Renders children if the client is authenticated.
 *
 * @public
 */
export function Authenticated({ children }: { children: ReactNode }) {
  const { isLoading, isAuthenticated } = useConvexAuth();
  if (isLoading || !isAuthenticated) {
    return null;
  }
  return <>{children}</>;
}

/**
 * Renders children if the client is using authentication but is not authenticated.
 * If `loadingEqualsUnauthenticated` is `true`, also renders children while the
 * client is authenticating.
 *
 * @public
 */
export function Unauthenticated({
  children,
  loadingEqualsUnauthenticated = false,
}: {
  children: ReactNode;
  loadingEqualsUnauthenticated?: boolean;
}) {
  const { isLoading, isAuthenticated } = useConvexAuth();
  if ((isLoading && !loadingEqualsUnauthenticated) || isAuthenticated) {
    return null;
  }
  return <>{children}</>;
}

/**
 * Renders children if the client isn't using authentication or is in the process
 * of authenticating.
 *
 * @public
 */
export function AuthLoading({ children }: { children: ReactNode }) {
  const { isLoading } = useConvexAuth();
  if (!isLoading) {
    return null;
  }
  return <>{children}</>;
}
