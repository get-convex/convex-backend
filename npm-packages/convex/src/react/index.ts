/**
 * Tools to integrate Convex into React applications.
 *
 * This module contains:
 * 1. {@link ConvexReactClient}, a client for using Convex in React.
 * 2. {@link ConvexProvider}, a component that stores this client in React context.
 * 3. {@link Authenticated}, {@link Unauthenticated} and {@link AuthLoading} helper auth components.
 * 4. Hooks {@link useQuery}, {@link useMutation}, {@link useAction} and more for accessing this
 *    client from your React components.
 *
 * ## Usage
 *
 * ### Creating the client
 *
 * ```typescript
 * import { ConvexReactClient } from "convex/react";
 *
 * // typically loaded from an environment variable
 * const address = "https://small-mouse-123.convex.cloud"
 * const convex = new ConvexReactClient(address);
 * ```
 *
 * ### Storing the client in React Context
 *
 * ```typescript
 * import { ConvexProvider } from "convex/react";
 *
 * <ConvexProvider client={convex}>
 *   <App />
 * </ConvexProvider>
 * ```
 *
 * ### Using the auth helpers
 *
 * ```typescript
 * import { Authenticated, Unauthenticated, AuthLoading } from "convex/react";
 *
 * <Authenticated>
 *   Logged in
 * </Authenticated>
 * <Unauthenticated>
 *   Logged out
 * </Unauthenticated>
 * <AuthLoading>
 *   Still loading
 * </AuthLoading>
 * ```
 *
 * ### Using React hooks
 *
 * ```typescript
 * import { useQuery, useMutation } from "convex/react";
 * import { api } from "../convex/_generated/api";
 *
 * function App() {
 *   const counter = useQuery(api.getCounter.default);
 *   const increment = useMutation(api.incrementCounter.default);
 *   // Your component here!
 * }
 * ```
 * @module
 */
export * from "./use_paginated_query.js";
export { useQueries, type RequestForQueries } from "./use_queries.js";
export type { AuthTokenFetcher, PlaintextAuthTokenFetcher } from "../browser/sync/client.js";
export * from "./auth_helpers.js";
export * from "./ConvexAuthState.js";
export * from "./hydration.js";
/* @internal */
export { useSubscription } from "./use_subscription.js";
export {
  type ReactMutation,
  type ReactAction,
  type Watch,
  type WatchQueryOptions,
  type MutationOptions,
  type ConvexReactClientOptions,
  type OptionalRestArgsOrSkip,
  ConvexReactClient,
  useConvex,
  ConvexProvider,
  useQuery,
  useMutation,
  useAction,
  useConvexConnectionState,
} from "./client.js";
