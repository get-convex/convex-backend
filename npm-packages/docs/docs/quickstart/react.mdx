---
title: React Quickstart
sidebar_label: React
description: "Add Convex to a React project"
hide_table_of_contents: true
sidebar_position: 100
---

import sampleData from "!!raw-loader!@site/../private-demos/quickstarts/react-vite/sampleData.jsonl";
import tasks from "!!raw-loader!@site/../private-demos/quickstarts/react-vite/convex/tasks.js";
import main from "!!raw-loader!@site/../private-demos/quickstarts/react-vite/src/main.jsx";
import App from "!!raw-loader!@site/../private-demos/quickstarts/react-vite/src/App.jsx";
import tasksTS from "!!raw-loader!@site/../private-demos/quickstarts/react-vite-ts/convex/tasks.ts";
import mainTS from "!!raw-loader!@site/../private-demos/quickstarts/react-vite-ts/src/main.tsx";
import AppTS from "!!raw-loader!@site/../private-demos/quickstarts/react-vite-ts/src/App.tsx";

To get setup quickly with Convex and React run

<p>
  <b>
    <CodeWithCopyButton text="npm create convex@latest" />
  </b>
</p>

or follow the guide below.

---

Learn how to query data from Convex in a React app using Vite
and<LanguageSelector verbose />

<StepByStep>
  <Step title="Create a React app">
    Create a React app using the `create vite` command.

    <JSDialectVariants>
      ```sh
      npm create vite@latest my-app -- --template react-ts
      ```

      ```sh
      npm create vite@latest my-app -- --template react
      ```
    </JSDialectVariants>

  </Step>
  <Step title="Install the Convex client and server library">
    To get started, install the `convex`
    package which provides a convenient interface for working
    with Convex from a React app.

    Navigate to your app directory and install `convex`.


    ```sh
    cd my-app && npm install convex
    ```

  </Step>
  <Step title="Set up a Convex dev deployment">
    Next, run `npx convex dev`. This
    will prompt you to log in with GitHub,
    create a project, and save your production and deployment URLs.

    It will also create a `convex/` folder for you
    to write your backend API functions in. The `dev` command
    will then continue running to sync your functions
    with your dev deployment in the cloud.


    ```sh
    npx convex dev
    ```

  </Step>

  <Step title="Create sample data for your database">
    In a new terminal window, create a `sampleData.jsonl`
    file with some sample data.

    <Snippet
      source={sampleData}
      title="sampleData.jsonl"
    />

  </Step>

  <Step title="Add the sample data to your database">
    Now that your project is ready, add a `tasks` table
    with the sample data into your Convex database with
    the `import` command.

    ```
    npx convex import --table tasks sampleData.jsonl
    ```

  </Step>

  <Step title="(optional) Define a schema">
    Add a new file `schema.ts` in the `convex/` folder
    with a description of your data.

    This will declare the types of your data for optional
    typechecking with TypeScript, and it will be also
    enforced at runtime.

    <JSDialectVariants>
    Alternatively remove the line `'plugin:@typescript-eslint/recommended-requiring-type-checking',`
    from the `.eslintrc.cjs` file to lower the type checking strictness.

    <></>
    </JSDialectVariants>

    ```ts noDialect title="convex/schema.ts"
    import { defineSchema, defineTable } from "convex/server";
    import { v } from "convex/values";

    export default defineSchema({
      tasks: defineTable({
        text: v.string(),
        isCompleted: v.boolean(),
      }),
    });
    ```

  </Step>

  <Step title="Expose a database query">
    Add a new file <JSDialectFileName name="tasks.ts" /> in the `convex/` folder
    with a query function that loads the data.

    Exporting a query function from this file
    declares an API function named after the file
    and the export name, `api.tasks.get`.

    <TSAndJSSnippet
      sourceTS={tasksTS}
      sourceJS={tasks}
      title="convex/tasks.ts"
    />

  </Step>

  <Step title="Connect the app to your backend">
    In <JSDialectFileName name="src/main.jsx" />, create a `ConvexReactClient` and pass it to a `ConvexProvider`
    wrapping your app.

    <TSAndJSSnippet
      sourceTS={mainTS}
      sourceJS={main}
      title="src/main.tsx"
      highlightPatterns={[ "Convex", ]}
    />

  </Step>

  <Step title="Display the data in your app">
    In <JSDialectFileName name="src/App.jsx" />, use the `useQuery` hook to fetch from your `api.tasks.get`
    API function and display the data.

    <TSAndJSSnippet
      sourceTS={AppTS}
      sourceJS={App}
      title="src/App.tsx"
      highlightPatterns={[ "useQuery", "api", "tasks", "text", "\\)\\)\\}" ]}
    />

  </Step>

  <Step title="Start the app">
    Start the app, open [http://localhost:5173/](http://localhost:5173/) in a browser,
    and see the list of tasks.

    ```sh
    npm run dev
    ```

  </Step>

</StepByStep>

Using `create-react-app`? See the
[Create React App](/docs/client/react/quickstart-react-cra.mdx) version of this
guide.
