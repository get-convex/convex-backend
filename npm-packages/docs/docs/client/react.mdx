---
title: "Convex React"
sidebar_label: "React"
sidebar_position: 4
---

import SkipBad from "!!raw-loader!@site/../private-demos/snippets/src/reactSkipBad.tsx";
import SkipGood from "!!raw-loader!@site/../private-demos/snippets/src/reactSkipGood.tsx";

Convex React is the client library enabling your React application to interact
with your Convex backend. It allows your frontend code to:

1. Call your [queries](/docs/functions/query-functions.mdx),
   [mutations](/docs/functions/mutation-functions.mdx) and
   [actions](/docs/functions/actions.mdx)
2. Upload and display files from [File Storage](/docs/file-storage.mdx)
3. Authenticate users using [Authentication](/docs/auth.mdx)
4. Implement full text [Search](/docs/search.mdx) over your data

The Convex React client is open source and available on
[GitHub](https://github.com/get-convex/convex-js).

Follow the [React Quickstart](/docs/quickstart/react.mdx) to get started with
React using [Vite](https://vitejs.dev/).

## Installation

Convex React is part of the `convex` npm package:

```
npm install convex
```

## Connecting to a backend

The [`ConvexReactClient`](/api/classes/react.ConvexReactClient) maintains a
connection to your Convex backend, and is used by the React hooks described
below to call your functions.

First you need to create an instance of the client by giving it your backend
deployment URL. See
[Configuring Deployment URL](/docs/client/react/project-setup.md) on how to pass
in the right value:

```jsx
import { ConvexProvider, ConvexReactClient } from "convex/react";

const convex = new ConvexReactClient("https://<your domain here>.convex.cloud");
```

And then you make the client available to your app by passing it in to a
[`ConvexProvider`](/api/modules/react#convexprovider) wrapping your component
tree:

```jsx
reactDOMRoot.render(
  <React.StrictMode>
    <ConvexProvider client={convex}>
      <App />
    </ConvexProvider>
  </React.StrictMode>,
);
```

## Fetching data

Your React app fetches data using the [`useQuery`](/api/modules/react#usequery)
React hook by calling your [queries](/docs/functions/query-functions.mdx) via an
[`api`](/docs/generated-api/api.md#api) object.

The `npx convex dev` command generates this api object for you in the
`convex/_generated/api.js` module to provide better autocompletion in JavaScript
and end-to-end type safety in
[TypeScript](/docs/understanding/best-practices/typescript.mdx):

```tsx title="src/App.tsx"
import { useQuery } from "convex/react";
import { api } from "../convex/_generated/api";

export function App() {
  const data = useQuery(api.functions.myQuery);
  return data ?? "Loading...";
}
```

The `useQuery` hook returns `undefined` while the data is first loading and
afterwards the return value of your query.

### Query arguments

Arguments to your query follow the query name:

```tsx title="src/App.tsx"
export function App() {
  const a = "Hello world";
  const b = 4;
  const data = useQuery(api.functions.myQuery, { a, b });
  //...
}
```

### Reactivity

The `useQuery` hook makes your app automatically reactive: when the underlying
data changes in your database, your component rerenders with the new query
result.

The first time the hook is used it creates a subscription to your backend for a
given query and any arguments you pass in. When your component unmounts, the
subscription is canceled.

### Consistency

Convex React ensures that your application always renders a consistent view of
the query results based on a single state of the underlying database.

Imagine a mutation changes some data in the database, and that 2 different
`useQuery` call sites rely on this data. Your app will never render in an
inconsistent state where only one of the `useQuery` call sites reflects the new
data.

### Paginating queries

See
[Paginating within React Components](/docs/database/pagination.mdx#paginating-within-react-components).

### Skipping queries

<Details summary="Advanced: Loading a query conditionally">

With React it can be tricky to dynamically invoke a hook, because hooks cannot
be placed inside conditionals or after early returns:

<TSAndJSSnippet
  title="src/App.tsx"
  sourceTS={SkipBad}
  sourceJS={SkipBad}
  snippet="example"
  highlightPatterns={["useQuery"]}
/>

For this reason `useQuery` can be "disabled" by passing in `"skip"` instead of
its arguments:

<TSAndJSSnippet
  title="src/App.tsx"
  sourceTS={SkipGood}
  sourceJS={SkipGood}
  highlightPatterns={["skip"]}
/>

When `"skip"` is used the `useQuery` doesn't talk to your backend at all and
returns `undefined`.

</Details>

### One-off queries

<Details summary="Advanced: Fetching a query from a callback">

Sometimes you might want to read state from the database in response to a user
action, for example to validate given input, without making any changes to the
database. In this case you can use a one-off
[`query`](/api/classes/react.ConvexReactClient#query) call, similarly to calling
mutations and actions.

The async method `query` is exposed on the `ConvexReactClient`, which you can
reference in your components via the
[`useConvex()`](/api/modules/react#useconvex) hook.

```tsx title="src/App.tsx"
import { useConvex } from "convex/react";
import { api } from "../convex/_generated/api";

export function App() {
  const convex = useConvex();
  return (
    <button
      onClick={async () => {
        console.log(await convex.query(api.functions.myQuery));
      }}
    >
      Check
    </button>
  );
}
```

</Details>

## Editing data

Your React app edits data using the
[`useMutation`](/api/modules/react#usemutation) React hook by calling your
[mutations](/docs/functions/mutation-functions.mdx).

The `convex dev` command generates this api object for you in the
`convex/_generated/api.js` module to provide better autocompletion in JavaScript
and end-to-end type safety in
[TypeScript](/docs/understanding/best-practices/typescript.mdx):

```tsx title="src/App.tsx"
import { useMutation } from "convex/react";
import { api } from "../convex/_generated/api";

export function App() {
  const doSomething = useMutation(api.functions.doSomething);
  return <button onClick={() => doSomething()}>Click me</button>;
}
```

The hook returns an `async` function which performs the call to the mutation.

### Mutation arguments

Arguments to your mutation are passed to the `async` function returned from
`useMutation`:

```tsx title="src/App.tsx"
export function App() {
  const a = "Hello world";
  const b = 4;
  const doSomething = useMutation(api.functions.doSomething);
  return <button onClick={() => doSomething({ a, b })}>Click me</button>;
}
```

### Mutation response and error handling

The mutation can optionally return a value or throw errors, which you can
[`await`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/await):

```tsx title="src/App.tsx"
export function App() {
  const doSomething = useMutation(api.functions.doSomething);
  const onClick = () => {
    async function callBackend() {
      try {
        const result = await doSomething();
      } catch (error) {
        console.error(error);
      }
      console.log(result);
    }
    void callBackend();
  };
  return <button onClick={onClick}>Click me</button>;
}
```

Or handle as a
[`Promise`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise):

```tsx title="src/App.tsx"
export function App() {
  const doSomething = useMutation(api.functions.doSomething);
  const onClick = () => {
    doSomething()
      .catch((error) => {
        console.error(error);
      })
      .then((result) => {
        console.log(result);
      });
  };
  return <button onClick={onClick}>Click me</button>;
}
```

Learn more about
[Error Handling](/docs/functions/error-handling/error-handling.mdx) in
functions.

### Retries

Convex React automatically retries mutations until they are confirmed to have
been written to the database. The Convex backend ensures that despite multiple
retries, every mutation call only executes once.

Additionally, Convex React will warn users if they try to close their browser
tab while there are outstanding mutations. This means that when you call a
Convex mutation, you can be sure that the user's edits won't be lost.

### Optimistic updates

Convex queries are fully reactive, so all query results will be automatically
updated after a mutation. Sometimes you may want to update the UI before the
mutation changes propagate back to the client. To accomplish this, you can
configure an _optimistic update_ to execute as part of your mutation.

Optimistic updates are temporary, local changes to your query results which are
used to make your app more responsive.

See [Optimistic Updates](/docs/client/react/optimistic-updates.mdx) on how to
configure them.

## Calling third-party APIs

Your React app can read data, call third-party services, and write data with a
single backend call using the [`useAction`](/api/modules/react#useaction) React
hook by calling your [actions](/docs/functions/actions.mdx).

Like `useQuery` and `useMutation`, this hook is used with the `api` object
generated for you in the `convex/_generated/api.js` module to provide better
autocompletion in JavaScript and end-to-end type safety in
[TypeScript](/docs/understanding/best-practices/typescript.mdx):

```tsx title="src/App.tsx"
import { useAction } from "convex/react";
import { api } from "../convex/_generated/api";

export function App() {
  const doSomeAction = useAction(api.functions.doSomeAction);
  return <button onClick={() => doSomeAction()}>Click me</button>;
}
```

The hook returns an `async` function which performs the call to the action.

### Action arguments

Action arguments work exactly the same as
[mutation arguments](#mutation-arguments).

### Action response and error handling

Action response and error handling work exactly the same as
[mutation response and error handling](#mutation-response-and-error-handling).

Actions do not support automatic retries or optimistic updates.

## Under the hood

The [`ConvexReactClient`](/api/classes/react.ConvexReactClient) connects to your
Convex deployment by creating a
[`WebSocket`](https://developer.mozilla.org/en-US/docs/Web/API/WebSocket). The
WebSocket provides a 2-way communication channel over TCP. This allows Convex to
push new query results reactively to the client without the client needing to
poll for updates.

If the internet connection drops, the client will handle reconnecting and
re-establishing the Convex session automatically.
