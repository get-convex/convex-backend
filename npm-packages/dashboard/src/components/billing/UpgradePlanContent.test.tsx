import {
  render,
  screen,
  fireEvent,
  renderHook,
  waitFor,
} from "@testing-library/react";
import { useFormik } from "formik";
import { act } from "react";
import { CreateSubscriptionArgs, PlanResponse } from "generatedApi";
import {
  CreateSubscriptionSchema,
  PriceInDollars,
  UpgradePlanContent,
  UpgradePlanContentProps,
} from "./UpgradePlanContent";

jest.mock("api/profile", () => {});
jest.mock("api/billing", () => {});
jest.mock("next/router", () => jest.requireActual("next-router-mock"));

describe("UpgradePlanContent", () => {
  const mockPlan: PlanResponse = {
    name: "Basic",
    id: "basic",
    planType: "basic",
    seatPrice: 10,
    description: "abc",
    status: "active",
  };
  const mockNumMembers = 5;
  const mockOnCreateSubscription = jest.fn();
  function newFormState(
    initialValues?: Partial<CreateSubscriptionArgs & { promoCode?: string }>,
  ) {
    return renderHook(() =>
      useFormik<CreateSubscriptionArgs & { promoCode?: string }>({
        initialValues: {
          promoCode: "",
          name: "",
          email: "",
          planId: "123",
          paymentMethod: undefined,
          ...initialValues,
        },
        validationSchema: CreateSubscriptionSchema,
        onSubmit: (v) => {
          mockOnCreateSubscription(v);
        },
      }),
    );
  }
  let mockFormState = newFormState();

  beforeEach(() => {
    jest.resetAllMocks();
    mockFormState = newFormState();
  });

  function renderUI(props?: Partial<UpgradePlanContentProps>) {
    render(
      <UpgradePlanContent
        formState={mockFormState.result.current}
        plan={mockPlan}
        numMembers={mockNumMembers}
        setPaymentMethod={jest.fn()}
        billingAddressInputs={null}
        paymentDetailsForm={null} // Add the required paymentDetailsForm property
        {...props}
      />,
    );
  }

  it("renders the price summary", async () => {
    renderUI();
    const priceSummaryElement = screen.getByTestId("price-summary");
    await waitFor(() => expect(priceSummaryElement).toBeInTheDocument());
  });

  it("does not render the payment method button when the plan is free", async () => {
    renderUI({ teamMemberDiscountPct: 1 });
    const paymentMethodElement = screen.queryByTestId(
      "update-payment-method-button",
    );
    await waitFor(() => expect(paymentMethodElement).not.toBeInTheDocument());
  });

  it("renders the promo code input", async () => {
    renderUI();
    const promoCodeInput = screen.getByLabelText("Promo code");
    await waitFor(() => expect(promoCodeInput).toBeInTheDocument());
  });

  it("renders the promo code spinner while loading", () => {
    renderUI({ isLoadingPromo: true });
    screen.getByTestId("loading-spinner");
  });

  it("renders the promo code error", () => {
    const promoCodeError = "Invalid promo code";
    renderUI({ promoCodeError });
    screen.getByText(promoCodeError);
  });

  it("calls onCreateSubscription when upgrade plan button is clicked", async () => {
    mockFormState = newFormState({
      promoCode: "",
      name: "Ari",
      email: "ari@convex.dev",
      planId: "123",
      paymentMethod: "wah",
      billingAddress: {
        line1: "123 Main St",
        line2: null,
        city: "San Francisco",
        state: "CA",
        postal_code: "94105",
        country: "US",
      },
    });
    renderUI({ formState: mockFormState.result.current });
    act(() => {
      const upgradePlanButton = screen.getByTestId("upgrade-plan-button");
      expect(upgradePlanButton).not.toBeDisabled();
      fireEvent.click(upgradePlanButton);
    });
    await waitFor(() => {
      expect(mockOnCreateSubscription).toHaveBeenCalledWith(
        mockFormState.result.current.values,
      );
    });
  });
});

describe("PriceInDollars", () => {
  it("renders the price", () => {
    const price = 100;
    const percentOff = 0;
    const { container } = render(
      <PriceInDollars price={price} percentOff={percentOff} />,
    );
    expect(container).toMatchSnapshot();
  });

  it("renders the price and percent off", () => {
    const price = 100;
    const percentOff = 0.2;
    const { container } = render(
      <PriceInDollars price={price} percentOff={percentOff} />,
    );
    expect(container).toMatchSnapshot();
  });
});
