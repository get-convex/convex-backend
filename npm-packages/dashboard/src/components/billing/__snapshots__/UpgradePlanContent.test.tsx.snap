// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PriceInDollars renders the price 1`] = `
<div>
  <span
    class="font-semibold"
  >
    $
    100
  </span>
</div>
`;

exports[`PriceInDollars renders the price and percent off 1`] = `
<div>
  <span
    class="mr-1 line-through"
  >
    $
    100
  </span>
  <span
    class="font-semibold"
  >
    $
    80
  </span>
</div>
`;
